README.md
# ngx_log_collection
This is a third part module for nginx to collect the logs that clients send, it can be used in versions 1.2.6, 1.8.1 and the newest official version 1.9.12, other versions have not been tested. It is compatible for standard HTTP POST method(Content-Type: x-www-form-urlencoded). Project also includes web testing files(JavaScript/jQuery) and client testing files(C++).


server/ChangeLog
For batch clients test log collection.

Version 1.0:
Happy road version.
First code on 2011-07-15.

Version 1.1:
Add cleanup.
Fix bugs.
Code on 2011-07-20.

Version 1.2:
Rename log file to a new one if it reaches the max size.
Add processing of Expect.
Code on 2011-07-22.

Version 1.3:
Fix bugs: add "ctx->main->count--;" in post_handler.
Code on 2011-07-25.

Version 1.4:
Add urldecode.
Fix bugs.
Code on 2011-07-26.

Version 1.5:
Add backend processing.
Code on 2011-07-29.

Version 1.6:
Fix a bug: parse log_collection, add "cf->args->nelts == 3".
Fix a bug: if redirect_to_backend is 0, do not run script.
Code on 2011-08-10.

Version 1.7:
Open or close urldecode depends on nginx.conf.
Code on 2011-08-12.

Version 2.0:
Add compatibility for versions above 1.3.9.
Code on 2016-03-28.

Fix a bug: initialize rc = NGX_OK in ngx_http_log_collection_post_handler.
Code on 2016-06-10.

Fix a bug: initialize rest = 0 in ngx_http_log_collection_flush_to_file.
Code on 2016-09-26.

Version 2.1:
Recode body reading functionality,
refering to https://github.com/hongzhidao/nginx-upload-module, many thanks.
Code on 2016-10-09.

Version 2.2:
Rename some functions and variables in order to be easily.
Add data purification.
Code on 2016-10-11.

Version 2.3:
Beautify the code.
Add backend processing depends on the upload content.
Code on 2016-10-15.

Version 2.4:
Fix a bug: if redirect to backend disabled, nothing stored.
Fix a bug: urldecode ' ' -> '+'
Code on 2016-10-16.

Add content length in normal response.
Optimize the code.
Code on 2016-10-17 & 2016-10-18.

Fix the compatibility problem.
Code on 2016-10-21.

Fix the bug that the garbage data in buffer will lead to wrongly parsing.
Add Changelog file.
Code on 2016-10-21.

Remove the upload functionality using chunk.
Code on 2017-01-13.

Version 2.5:
Optimize the writing log functionality.
Fix some bugs.
Code on 2017-02-03.

Optimize the code.
Code on 2017-02-04.

Version 2.6:
Fix a bug: when finishing the request, subtract the connection's requests
member, in order not to reach the max count of the pipeline requests on a
connection(if so, nginx will close the connection by force).
Optimize the code.
Code on 2017-02-22.

Version 2.7:
Add a new branch and in which add UUID authentication functionality.
Code on 2017-02-27.

Version 2.8:
Add a new functionality that supports pipeline processing.
Recode the C++ testing client.
Code on 2017-03-02.

Fix a bug: when compiled with parameter "--with-debug", it will crash when
writing the received log to the file.
Fix a bug: add the unlock when processing a wrong case.
Add a wrong form data processing.
Code on 2017-05-06.



client/src/main.cpp
#include "http_socket.h"
#include <vector>
#include <cstdlib>

using std::vector;

int main(int argc, char* argv[])
{
	/* default */
	unsigned int worker = 5, loop = 1000;
	unsigned short port = 80;
	stringstream s;
	const string uuid_prefix = "c20aff8a-c457-42b1-a096-";
	string uuid, server, h = "-h", help = "--help";
	CHttpSocket* test = NULL;
	vector<CHttpSocket*> test_array;
	
	if (argc == 2 && (h == argv[1] || help == argv[1])) {
		cout << "Usage: " << argv[0] << " [server] [#port] [#worker] [#loop]" << endl;
		cout << "Default: " << argv[0] << " localhost 80 5 1000" << endl;
		exit(-1);
	} else if (argc == 1) {
		server = "localhost";
	} else if (argc == 2) {
		server = argv[1];
	} else if (argc == 3) {
		server = argv[1];

		s << argv[2];
		s >> port;
	} else if (argc == 4) {
		server = argv[1];

		s << argv[2];
		s >> port;
		s.clear();

		s << argv[3];
		s >> worker;
	} else {
		server = argv[1];

		s << argv[2];
		s >> port;
		s.clear();

		s << argv[3];
		s >> worker;
		s.clear();

		s << argv[4];
		s >> loop;
		s.clear();
	}

	if (worker <= 0 || worker > 500) {
		cout << "Limit worker to 500" << endl; 
		worker = 500;
	}

	test_array.reserve(worker);

	for (unsigned int i = 0; i < worker; ++i) {
		s.clear();
		s.str("");
		uuid.clear();

		s << i;
		s >> uuid;

		string::size_type size = uuid.size();
		for (unsigned int j = 0; j < 12 - size; ++j) {
			uuid = "0" + uuid;
		}

		uuid = uuid_prefix + uuid;
		test = new CHttpSocket(uuid, server, port, "/log_collection", loop);

		if (test) {
			test_array.push_back(test);
			test_array[i]->StartWork();
		}
	}

	for (unsigned int i = 0; i < test_array.size(); ++i) {
		delete test_array[i];
	}

	return 0;
}



client/include/http_socket.h
#include <cstdio>
#include <sys/epoll.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <pthread.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <netdb.h>
#include <signal.h>
#include <arpa/inet.h>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>

using std::cout;
using std::endl;
using std::string;
using std::stringstream;

typedef enum {
	RESPONSE_START,
	RESPONSE_HEADER,
	RESPONSE_INVALID,
	RESPONSE_DONE
} RespState;

class CHttpSocket
{
public:
	CHttpSocket(const string& uuid, const string& host = "localhost",
		const unsigned short port = 80, const string& uri = "/",
		const int loop = -1, const int timeout = 50);
	~CHttpSocket();

	void	StartWork(void);
	const int& GetConnectFd(void) const;
	const int& GetEpollFd(void) const;
	const string& GetReqHeader(void) const;
	const int& GetTimeout(void) const;
	const int& GetLoop(void) const;
	int		SetNonBlockConnect(void);
	bool	CheckConnectError(void);

protected:
	size_t	FormatRequestHeader(void);
	void	Initialize(const string& host, const unsigned short port,
				const string& uri, const int loop, const int timeout);
	bool	ReInitialize(struct epoll_event& ev);
	void	Finalize(void);

private:
	bool	ResolveServer(struct sockaddr *serv, size_t *servlen);
	RespState	CheckResponse(const string& buffer);
	static	void* DoWork(void *arg);

	string	m_req;

	string	m_host;
	string	m_port;
	string	m_uri;
	string	m_uuid;

	int		m_timeout;

	int		m_epollfd;
	int		m_connfd;
	int		m_worker;
	int		m_loop;
	int		m_pipeline;
	int		m_hdrend;

	pthread_t	m_tid;

	bool	m_inited;
	static	const int s_bufsize;
};



server/ngx_http_uuid_authen.c
/*
 * Copyright (C) 2017 winshining https://github.com/winshining
 * Copyright (C) 2002-2017 Igor Sysoev
 * Copyright (C) 2011-2017 Nginx, Inc.
 */

#include "ngx_http_log_collection_module.h"

ngx_int_t
ngx_http_uuid_authen_shm_slab_init(ngx_shm_zone_t *shm_zone, void *data)
{
	size_t len;
	ngx_http_log_collection_loc_conf_t *conf;
	ngx_http_log_collection_loc_conf_t *oldconf;

	conf = (ngx_http_log_collection_loc_conf_t *) shm_zone->data;
	oldconf = (ngx_http_log_collection_loc_conf_t *) data;

	if (oldconf) {
		conf->uuid_authen_conf.sh = oldconf->uuid_authen_conf.sh;
		conf->uuid_authen_conf.shpool = oldconf->uuid_authen_conf.shpool;
		return NGX_OK;
	}

	conf->uuid_authen_conf.shpool = (ngx_slab_pool_t *)shm_zone->shm.addr;
	conf->uuid_authen_conf.sh = ngx_slab_alloc(conf->uuid_authen_conf.shpool,
			sizeof(ngx_http_uuid_authen_shm_slab_t));
	if (conf->uuid_authen_conf.sh == NULL) {
		return NGX_ERROR;
	}

	conf->uuid_authen_conf.shpool->data = conf->uuid_authen_conf.sh;

	/* initialize the red-black tree */
	ngx_rbtree_init(&conf->uuid_authen_conf.sh->rbtree, &conf->uuid_authen_conf.sh->sentinel,
			ngx_str_rbtree_insert_value);

	len = ngx_strlen(" in uuid authen slab \"\"") + shm_zone->shm.name.len;
	conf->uuid_authen_conf.shpool->log_ctx = ngx_slab_alloc(conf->uuid_authen_conf.shpool, len);
	if (conf->uuid_authen_conf.shpool->log_ctx == NULL) {
		return NGX_ERROR;
	}

	ngx_sprintf(conf->uuid_authen_conf.shpool->log_ctx, " in uuid authen slab \"%V\"%Z",
			&shm_zone->shm.name);

	return NGX_OK;
}

char *
ngx_http_uuid_authen_create_shm_slab(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
{
	ngx_str_t		*value, *name;
	ngx_shm_zone_t	*shm_zone;
	ngx_http_log_collection_loc_conf_t *lclcf;

	lclcf = (ngx_http_log_collection_loc_conf_t *)conf;
	lclcf->uuid_authen_conf.authen_switch = 0;

	value = cf->args->elts;
	name = &value[1];

	lclcf->uuid_authen_conf.shm_size = ngx_parse_size(&value[2]);
	if (lclcf->uuid_authen_conf.shm_size == (size_t) NGX_ERROR
		|| lclcf->uuid_authen_conf.shm_size == 0)
	{
		return "invalid value";
	}

	if (lclcf->uuid_authen_conf.shm_size < 10 * (1 << 20)) {
		lclcf->uuid_authen_conf.shm_size = 10 * (1 << 20);
	}

	shm_zone = ngx_shared_memory_add(cf, name, lclcf->uuid_authen_conf.shm_size,
			&ngx_http_log_collection_module);

	if (shm_zone == NULL) {
		lclcf->uuid_authen_conf.shm_size = 0;
		return NGX_CONF_ERROR;
	}

	shm_zone->init = ngx_http_uuid_authen_shm_slab_init;
	shm_zone->data = &lclcf->uuid_authen_conf;

	lclcf->uuid_authen_conf.authen_switch = 1;

	return NGX_CONF_OK;
}

ngx_int_t
ngx_http_uuid_authen_handler(ngx_http_request_t *r, ngx_str_t *uuid)
{
	uint32_t		hash;
	size_t			len;
	ngx_int_t		rc;
	ngx_uint_t		i;
	ngx_list_part_t	*part;
	ngx_table_elt_t	*header;
	ngx_str_node_t	*node;
	ngx_http_uuid_authen_node_t			*new;
	ngx_http_log_collection_loc_conf_t	*conf;

	conf = ngx_http_get_module_loc_conf(r, ngx_http_log_collection_module);

	if (r == NULL || uuid == NULL) {
		return NGX_HTTP_INTERNAL_SERVER_ERROR;
	}

	part = &r->headers_in.headers.part;
	header = part->elts;

	for (i = 0; /* void */; i++) {
		if (i >= part->nelts) {
			if (part->next == NULL) {
				break;
			}

			part = part->next;
			header = part->elts;
			i = 0;
		}

		if (header[i].hash == 0) {
			continue;
		}

		if (ngx_strncasecmp(header[i].key.data,	(u_char *) CLIENT_UUID_STRING,
			ngx_strlen(CLIENT_UUID_STRING)) == 0)
		{
			uuid->data = header[i].value.data;
			uuid->len = header[i].value.len;

			if (uuid->len != UUID_STRING_LENGTH
				|| uuid->data[8] != '-'
				|| uuid->data[13] != '-'
				|| uuid->data[18] != '-'
				|| uuid->data[23] != '-')
			{
				ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "Invalid UUID format: %V", uuid);
				return NGX_HTTP_FORBIDDEN;
			}

			break;
		}
	}

	if (uuid->len == 0) {
		/* uuid not found */
		return NGX_HTTP_BAD_REQUEST;
	}

	if (conf->uuid_authen_conf.authen_switch == 0) {
		return NGX_DECLINED;
	}

	hash = ngx_crc32_short(uuid->data, uuid->len);

	ngx_shmtx_lock(&conf->uuid_authen_conf.shpool->mutex);

	node = ngx_str_rbtree_lookup(&conf->uuid_authen_conf.sh->rbtree, uuid, hash);
	if (node) {
		if (r->pipeline) {
			ngx_log_error(NGX_LOG_WARN, r->connection->log, 0, "It is a pipelined request");

			rc = NGX_DECLINED;
		} else {
			ngx_log_error(NGX_LOG_ERR, r->connection->log, 0, "An uuid \"%V\" exists", uuid);

			rc = NGX_HTTP_FORBIDDEN;
		}
	} else {
		/* insert the uuid into the rbtree */
		len = sizeof(ngx_str_node_t) + UUID_STRING_LENGTH - 1;
		new = ngx_slab_alloc_locked(conf->uuid_authen_conf.shpool, len);
		if (new == NULL) {
			ngx_log_error(NGX_LOG_ERR, r->connection->log, 0,
					"Allocation for the node of uuid \"%V\" failed", uuid);

			rc = NGX_HTTP_FORBIDDEN;
		} else {
			ngx_memcpy(new->data, uuid->data, UUID_STRING_LENGTH);
			new->str_node.node.key = hash;
			new->str_node.str.data = new->data;
			new->str_node.str.len = ngx_strlen(new->data);
			ngx_rbtree_insert(&conf->uuid_authen_conf.sh->rbtree, &new->str_node.node);

			rc = NGX_DECLINED;
		}
	}

	ngx_shmtx_unlock(&conf->uuid_authen_conf.shpool->mutex);

	return rc;
}

void
ngx_http_uuid_authen_expire(ngx_http_request_t *r, ngx_str_t *uuid)
{
	uint32_t		hash;
	ngx_str_node_t	*node;
	ngx_http_log_collection_loc_conf_t	*lclcf;

	if (r == NULL || uuid == NULL) {
		return;
	}

	lclcf = ngx_http_get_module_loc_conf(r, ngx_http_log_collection_module);

	if (lclcf->uuid_authen_conf.authen_switch) {
		ngx_shmtx_lock(&lclcf->uuid_authen_conf.shpool->mutex);

		hash = ngx_crc32_short(uuid->data, uuid->len);
		node = ngx_str_rbtree_lookup(&lclcf->uuid_authen_conf.sh->rbtree, uuid, hash);
		if (node) {
			ngx_rbtree_delete(&lclcf->uuid_authen_conf.sh->rbtree, &node->node);
			ngx_slab_free_locked(lclcf->uuid_authen_conf.shpool, (void *) node);
		}

		ngx_shmtx_unlock(&lclcf->uuid_authen_conf.shpool->mutex);
	}
}



server/ngx_http_uuid_authen.h
/*
 * Copyright (C) 2016-2017 winshining https://github.com/winshining
 * Copyright (C) 2002-2017 Igor Sysoev
 * Copyright (C) 2011-2017 Nginx, Inc.
 */

#ifndef __NGX_HTTP_UUID_AUTHEN_H__
#define __NGX_HTTP_UUID_AUTHEN_H__

/* the data in ngx_str_t points to data,
 * force the node all locate in the shm
 */
typedef struct {
	ngx_str_node_t	str_node;
	u_char			data[1];
} ngx_http_uuid_authen_node_t;

typedef struct {
	ngx_rbtree_t		rbtree;
	ngx_rbtree_node_t	sentinel;
} ngx_http_uuid_authen_shm_slab_t;

typedef struct {
	ngx_flag_t						authen_switch;
	size_t							shm_size;

	/* located in the shared memory */
	ngx_slab_pool_t					*shpool;
	ngx_http_uuid_authen_shm_slab_t	*sh;
} ngx_http_uuid_authen_conf_t;

extern ngx_module_t ngx_http_log_collection_module;

ngx_int_t
ngx_http_uuid_authen_shm_slab_init(ngx_shm_zone_t *shm_zone, void *data);

ngx_int_t
ngx_http_uuid_authen_handler(ngx_http_request_t *r, ngx_str_t *uuid);

char *
ngx_http_uuid_authen_create_shm_slab(ngx_conf_t *cf, ngx_command_t *cmd, void *conf);

void
ngx_http_uuid_authen_expire(ngx_http_request_t *r, ngx_str_t *uuid);

#endif



server/ngx_http_log_collection_module.h
/*
 * Copyright (C) 2016 hongzhidao https://github.com/hongzhidao
 * Copyright (C) 2016-2017 winshining https://github.com/winshining
 * Copyright (C) 2006-2008 Valery Kholodkov
 * Copyright (C) 2002-2017 Igor Sysoev
 * Copyright (C) 2011-2017 Nginx, Inc.
 */

#ifndef __NGX_HTTP_LOG_COLLECTION_MODULE_H__
#define __NGX_HTTP_LOG_COLLECTION_MODULE_H__

#include <ngx_config.h>
#include <ngx_core.h>
#include <ngx_http.h>
#include "ngx_http_uuid_authen.h"

#define CLIENT_UUID_STRING			"Client-UUID" /* user defined */
#define UUID_STRING_LENGTH			36

/*
 * Log collection configuration for specific location
 */
typedef struct ngx_http_log_collection_loc_conf_s {
	/* for uuid authentication */
	ngx_http_uuid_authen_conf_t uuid_authen_conf;

	/* for backend */
	ngx_str_t	url;
	ngx_http_complex_value_t *url_cv;

	ngx_flag_t	all_in_single_folder;
	ngx_flag_t	log_collection_switch;
	ngx_flag_t	log_content_decode;
	ngx_flag_t	log_content_purify;
	ngx_flag_t	redirect_to_backend;
	ngx_path_t	*store_path;
	ngx_str_t	store_map_to_uri;

	ngx_array_t	*field_templates; /* ngx_http_log_collection_field_template_t */
	ngx_flag_t	forward_args;

	off_t		max_file_size;
	size_t		max_request_body_size;
	size_t		upload_limit_rate;
} ngx_http_log_collection_loc_conf_t;

#endif



client/src/http_socket.cpp
#include "http_socket.h"

const int CHttpSocket::s_bufsize = 2048;

CHttpSocket::CHttpSocket(const string& uuid, const string& host,
		const unsigned short port, const string& uri,
		const int loop, const int timeout)
{
	Initialize(host, port, uri, loop, timeout);

	if (uuid.empty()) {
		Finalize();
		cout << pthread_self() << ": empty uuid in construction" << endl;
	} else {
		m_uuid = uuid;
		m_inited = true;
	}
}

CHttpSocket::~CHttpSocket()
{
	Finalize();
}

size_t CHttpSocket::FormatRequestHeader(void)
{
	m_req.clear();
	m_req += "POST ";
	m_req += m_uri;
	m_req += " HTTP/1.1";
	m_req += "\r\n";

	m_req += "Host: ";
	m_req += m_host;
	m_req += "\r\n";

	if (m_uuid.size() != 0) {
		m_req += "Client-UUID: ";
		m_req += m_uuid;
		m_req += "\r\n";
	}

	m_req += "Accept: */*";
	m_req += "\r\n";

	m_req += "Connection: Keep-Alive";
	m_req += "\r\n";

	m_req += "Content-Type: application/x-www-form-urlencoded";
	m_req += "\r\n";

	m_req += "Content-Length: ";

	return (size_t) m_req.size();
}

void CHttpSocket::StartWork()
{
	int ret;

	if (!m_inited) {
		return;
	}

	FormatRequestHeader();

	ret = pthread_create(&m_tid, NULL, CHttpSocket::DoWork, (void *)this);
	if (ret != 0) {
		cout << pthread_self() << ": create thread failed:" << strerror(errno) << endl;
		return;
	}

	m_hdrend = 0;
	m_pipeline = 0;
	cout << m_tid << " is running" << endl;
}

RespState CHttpSocket::CheckResponse(const string& buffer)
{
	string::size_type pos = 0;
	string size;
	stringstream ss;
	RespState s = RESPONSE_START;

	if (buffer.empty()) {
		return (m_hdrend > 0) ? RESPONSE_START : RESPONSE_HEADER;
	}

	for ( ;; ) {
		if (m_hdrend > 0) {
			for ( ;; ) {
				/* Content-Length */
				pos = buffer.find("Content-Length:", pos);
				if (pos == string::npos) {
					if (m_pipeline == 0) {
						m_hdrend = 0;
						pos = 0;
						break;
					} else {
						return RESPONSE_HEADER; 
					}
				} else {
					m_pipeline++;
					if (m_pipeline > 2) {
						m_hdrend = 0;
						m_pipeline = 0;
						return RESPONSE_DONE;
					}
				}
			} 

			for ( ;; ) {
				/* Transfer-Encoding */
				pos = buffer.find("Transfer-Encoding:", pos);

				if (pos == string::npos) {
					if (m_pipeline == 0) {
						m_hdrend = 0;
						return RESPONSE_INVALID;
					} else {
						return RESPONSE_HEADER;
					}
				} else {
					m_pipeline++;
					if (m_pipeline > 2) {
						m_pipeline = 0;
						m_hdrend = 0;
						return RESPONSE_DONE;
					}
				}
			}
		} else {
			pos = buffer.find("\r\n\r\n", pos);

			if (pos != string::npos) {
				m_hdrend = (size_t) pos;
				s = RESPONSE_HEADER;
			} else {
				break;
			}
		}
	}

	return s;
}

void* CHttpSocket::DoWork(void *arg)
{
	const string str = "content=This+is+a+test+string+in+the+log+collection+client"
		"+programm%2C+first+coded+on+2016.03.08+by+Xingyuan+Wang%2C"
		"+which+is+used+to+test+the+log+collection+server+module+in+"
		"nginx,+versions+1.2.6,+1.8.1+and+1.9.12+were+tested+on+2016"
		"-10-10-15.%0D%0AThe+following+blanks+is+intended+to+tested+"
		"the+urldecode+functionality+of+the+module:           %0D%0A";

	stringstream s;
	string size, req;
	ssize_t all, r = 0, delta_in = 0, delta_out = 0;
	int n, i, j;
	char buffer[s_bufsize];
	struct epoll_event ev, events[10];
	bool connected = false, done;
	pthread_t tid = pthread_self();

	CHttpSocket* obj = (CHttpSocket*) arg;
	const string& reqh = obj->GetReqHeader();

	s << str.size();
	s >> size;
	req = reqh + size + "\r\n\r\n";
	req += str;
	req += reqh + size + "\r\n\r\n";
	req += str;
	req += reqh + size + "\r\n\r\n";
	req += str;
	all = req.size();

	signal(SIGPIPE, SIG_IGN);

	r = obj->SetNonBlockConnect();
	if (r == -1 && errno != EINPROGRESS) {
		cout << "Set nonblock connect failed." << endl;
		return NULL;
	}

	const int& connfd = obj->GetConnectFd();
	const int& epollfd = obj->GetEpollFd();
	const int& timeout = obj->GetTimeout();

	ev.events = EPOLLOUT | EPOLLET;
	ev.data.fd = connfd;
	n = epoll_ctl(epollfd, EPOLL_CTL_ADD, connfd, &ev);
	if (n == -1) {
		cout << tid << ": Add the connect socket for write failed: " << strerror(errno) << endl;
		return NULL;
	}

	for (i = 0; i < obj->GetLoop(); i++) {
		done = false;

		for ( ;; ) {
			n = epoll_wait(epollfd, events, 10, timeout);
			if (n == -1) {
				if (errno != EINTR) {
					cout << tid << ": Epoll failed: " << strerror(errno) << endl;
					return NULL;
				}
			} else if (n > 0) {
				for (j = 0; j < n; j++) {
					/* only one fd, so we do not check it */
					if (!connected) {
						if (!obj->CheckConnectError()) {
							cout << tid << ": Connect error happened." << endl;
							return NULL;
						} else {
							ev.events |= EPOLLIN;
							n = epoll_ctl(epollfd, EPOLL_CTL_MOD, connfd, &ev);
							if (n == -1 && errno != EINTR) {
								cout << tid << ": Add the connect socket for read failed: " <<
									strerror(errno) << endl;

								return NULL;
							}

							connected = true;
						}
					} else {
						if (events[j].events & EPOLLOUT) {
							/* send */
							while (delta_out < all) {
								r = send(connfd, req.c_str() + delta_out, req.size() - delta_out, 0);
								if (r == -1) {
									if (errno != EINTR && errno != EAGAIN) {
										cout << tid << ": Send error: " << strerror(errno) << endl;
										return NULL;
									}

									if (errno == EAGAIN) {
										delta_out = 0;
										break;
									}
								} else {
									if (delta_out >= all) {
										delta_out = 0;
										break;
									}

									delta_out += r;
								}
							}
						}

						if (events[j].events & EPOLLIN) {
							/* recv */
							for ( ;; ) {
								r = recv(connfd, buffer + delta_in, s_bufsize - delta_in, 0);
								if (r == -1) {
									if (errno != EINTR && errno != EAGAIN) {
										cout << tid << ": Recv error: " << strerror(errno) << endl;
										return NULL;
									} else {
										break;
									}
								} else if (r == 0) {
									delta_in = 0;
									cout << tid << ": Peer closed the connection." << endl;
									if (!obj->ReInitialize(ev)) {
										return NULL;
									}

									connected = false;
									break;
								} else {
									delta_in += r;

									RespState s = obj->CheckResponse(string(buffer));
									if (s == RESPONSE_DONE) {
										delta_in = 0;
										done = true;
										break;
									}
								}
							}
						}
					}
				}
			} else {
				if (done) {
					if (i == obj->GetLoop() - 1) {
						break;
					}

					delta_out = 0;
					while (delta_out < all) {
						r = send(connfd, req.c_str() + delta_out, req.size() - delta_out, 0);
						if (r == -1) {
							if (errno != EINTR && errno != EAGAIN) {
								cout << tid << ": Send error: " << strerror(errno) << endl;
								return NULL;
							}

							if (errno == EAGAIN) {
								continue;
							}
						} else {
							if (delta_out >= all) {
								delta_out = 0;
								break;
							}

							delta_out += r;
						}
					}

					break;
				}

				cout << tid << ": Epoll timeout." << endl;
			}
		}
	}

	return NULL;
}

void CHttpSocket::Initialize(const string& host, const unsigned short port,
	const string& uri, const int loop, const int timeout)
{
	m_connfd = -1;
	m_inited = false;
	m_tid = 0;

	m_host = host.empty() ? "localhost" : host;
	stringstream s;
	s << port;
	s >> m_port;
	m_uri = uri.empty() ? "/" : uri;
	m_loop = (loop <= 0) ? 50 : loop;
	m_timeout = (timeout <= 0) ? 50 : timeout;

	m_epollfd = epoll_create(10);
	if (m_epollfd == -1) {
		Finalize();
	}
}

bool CHttpSocket::ReInitialize(struct epoll_event& ev)
{
	int r, n;
	pthread_t tid = pthread_self();

	epoll_ctl(m_epollfd, EPOLL_CTL_DEL, m_connfd, NULL);
	close(m_connfd);
	m_connfd = -1;

	r = SetNonBlockConnect();
	if (r == -1 && errno != EINPROGRESS) {
		cout << "Set nonblock connect failed." << endl;
		return false;
	}

	ev.events = EPOLLOUT | EPOLLET;
	ev.data.fd = m_connfd;

	n = epoll_ctl(m_epollfd, EPOLL_CTL_ADD, m_connfd, &ev);
	if (n == -1) {
		cout << tid << ": Add the connect socket for write failed: " << strerror(errno) << endl;
		return false;
	}

	return true;
}

void CHttpSocket::Finalize(void)
{
	if (m_tid > 0) {
		pthread_join(m_tid, NULL);
		cout << m_tid << " joined" << endl;
	}

	if (m_epollfd >= 0) {
		if (m_connfd >= 0) {
			epoll_ctl(m_epollfd, EPOLL_CTL_DEL, m_connfd, NULL);
		}

		close(m_epollfd);
		m_epollfd = -1;
	}

	if (m_connfd >= 0) {
		close(m_connfd);
		m_connfd = -1;
	}

	m_inited = false;
	m_loop = 0;
}

bool CHttpSocket::ResolveServer(struct sockaddr *serv, size_t *servlen)
{
	int r;
	struct addrinfo hints, *res, *ressave;

	bzero(&hints, sizeof(hints));
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_family = AF_INET;

	r = getaddrinfo(m_host.c_str(), m_port.c_str(), &hints, &res);
	if (r != 0) {
		cout << pthread_self() << ": getaddrinfo error: " << gai_strerror(r) << endl;
		return false;
	}

	ressave = res;

	do {
		m_connfd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
		if (m_connfd > 0) {
			*servlen = res->ai_addrlen;
			bcopy(res->ai_addr, serv, *servlen);
			break;
		}
	} while((res = res->ai_next) != NULL);

	freeaddrinfo(ressave);

	if (res == NULL) {
		/* found nothing */
		cout << pthread_self() << ": " << m_host << " " << m_port << " not found" << endl;
		return false;
	}

	return true;
}

int CHttpSocket::SetNonBlockConnect(void)
{
	struct sockaddr srv;
	size_t addrlen;
	int	r = -1;

	if (ResolveServer(&srv, &addrlen)) {
		int val = fcntl(m_connfd, F_GETFL, 0);
		if (fcntl(m_connfd, F_SETFL, val | O_NONBLOCK) != -1) {
			r = connect(m_connfd, &srv, addrlen);
		}
	}

	return r;
}

bool CHttpSocket::CheckConnectError(void)
{
	int s = 1;
	socklen_t len = sizeof(s);

	if (getsockopt(m_connfd, SOL_SOCKET, SO_ERROR, (void *) &s, &len) == -1) {
		return false;
	} else {
		if (s != 0) {
			cout << pthread_self() << ": getsockopt for SO_ERROR failed." << endl;
			return false;
		}
	}

	return true;
}

const int& CHttpSocket::GetConnectFd(void) const
{
	return m_connfd;
}

const int& CHttpSocket::GetEpollFd(void) const
{
	return m_epollfd;
}

const string& CHttpSocket::GetReqHeader(void) const
{
	return m_req;
}

const int& CHttpSocket::GetTimeout(void) const
{
	return m_timeout;
}

const int& CHttpSocket::GetLoop(void) const
{
	return m_loop;
}

